File: content/examples/other/leetcode/problems/serialize-and-deserialize-binary-tree/example-2/solution.js
class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function serializeDFS(root) {
  const out = [];
  (function dfs(node) {
    if (!node) {
      out.push("#");
      return;
    }
    out.push(String(node.val));
    dfs(node.left);
    dfs(node.right);
  })(root);
  return out.join(",");
}

function deserializeDFS(data) {
  if (!data) return null;
  const parts = data.split(",");
  let i = 0;
  function build() {
    if (i >= parts.length) return null;
    const token = parts[i++];
    if (token === "#") return null;
    const node = new TreeNode(Number(token));
    node.left = build();
    node.right = build();
    return node;
  }
  return build();
}

if (require.main === module) {
  const root = new TreeNode(1, new TreeNode(2), new TreeNode(3, new TreeNode(4), new TreeNode(5)));
  const s = serializeDFS(root);
  console.log(s);
  console.log(serializeDFS(deserializeDFS(s)));
}

module.exports = { TreeNode, serializeDFS, deserializeDFS };

File: content/examples/other/leetcode/problems/serialize-and-deserialize-binary-tree/example-2/solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def serialize_dfs(root):
    out = []

    def dfs(node):
        if node is None:
            out.append("#")
            return
        out.append(str(node.val))
        dfs(node.left)
        dfs(node.right)

    dfs(root)
    return ",".join(out)


def deserialize_dfs(data):
    if not data:
        return None
    parts = data.split(",")
    i = 0

    def build():
        nonlocal i
        token = parts[i]
        i += 1
        if token == "#":
            return None
        node = TreeNode(int(token))
        node.left = build()
        node.right = build()
        return node

    return build()


if __name__ == "__main__":
    root = TreeNode(1, TreeNode(2), TreeNode(3, TreeNode(4), TreeNode(5)))
    s = serialize_dfs(root)
    print(s)
    print(serialize_dfs(deserialize_dfs(s)))
